app/classes/stock.rb

```diff
@@ -0,0 +1,167 @@
+        .where('tilausrivi.kerayspvm <= ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      stock_reserved += product.stock_transfer_rows
+        .where('tilausrivi.kerayspvm <= ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      # sales, manufacture, and stock trasfer rows
+      # *reserve stock* if they are due to be picked in the future, but are already picked
+      stock_reserved += product.sales_order_rows
+        .picked
+        .where('tilausrivi.kerayspvm > ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      stock_reserved += product.manufacture_rows
+        .picked
+        .where('tilausrivi.kerayspvm > ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      stock_reserved += product.stock_transfer_rows
+        .picked
+        .where('tilausrivi.kerayspvm > ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      # manufacture composite rows and manufacture recursive composite rows
+      # *decrease stock reservation* if they are due to be picked in the past
+      stock_reserved -= product.manufacture_composite_rows
+        .where('tilausrivi.kerayspvm <= ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      stock_reserved -= product.manufacture_recursive_composite_rows
+        .where('tilausrivi.kerayspvm <= ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      # purchase orders due to arrive in the past *decrease stock reservation*
+      stock_reserved -= product.purchase_order_rows
+        .where('tilausrivi.toimaika <= ?', stock_date)
+        .send(*@warehouse_filter)
+        .reserved
+
+      stock_reserved
+    end
+
+    def pick_date_and_future_reserved
+      relations = %w(
+        manufacture_composite_rows
+        manufacture_recursive_composite_rows
+        manufacture_rows
+        purchase_order_rows
+        sales_order_rows
+        stock_transfer_rows
+      )
+
+      # fetch all distinct pick dates for all product rows
+      dates = [stock_date]
+      dates << relations.map do |relation|
+        product.send(relation)
+          .where('tilausrivi.kerayspvm > ?', stock_date)
+          .where('tilausrivi.varattu + tilausrivi.jt != 0')
+          .send(*@warehouse_filter)
+          .select(:kerayspvm)
+          .distinct
+          .map(&:kerayspvm)
+      end
+
+      # fetch stock reserved for each date
+      stock_by_date = dates.flatten.compact.map do |date|
+        pick_date_stock_reserved(stock_date: date)
+      end
+
+      # return maximum stock reservation in the future (worst case)
+      stock_by_date.max || 0
+    end
+
+    def per_warehouse(method)
+      old_filter = @warehouse_filter
+
+      warehouse_ids.each_with_object({}) do |warehouse_id, stocks|
+        @warehouse_filter = [:where, varasto: warehouse_id]
+
+        stocks[warehouse_id] = send(method)
```

Comment:
En oikein diggaa tosta instanssimuuttujakikasta. Mutta ehkä sen kanssa voidaan mennä eteenpäin.

Mutta eikö `map` olisi tässä selkeämpi?

```ruby
warehouse_ids.map do |warehouse_id|
  @warehouse_filter = [:where, varasto: warehouse_id]

  { warehouse_id => send(method) }
end